<a href="https://colab.research.google.com/github/yilinw762/ds2002-fa26/blob/main/notebooks/04-json-apis/2026_10_07_%E2%80%94_APIs_101_and_Ingestion_with_Retries_%E2%80%94_Studio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DS2002 · APIs 101 and Ingestion with Retries

**Studio — 2026-10-07 · Fall 2026**  
**Class time:** 45 minutes

---

## Part 1 — What an API is

Every dataset you have opened this semester sat still. A CSV of Chipotle orders is the same file today as it was last week, because somebody exported it once and the export stopped being true the moment it was written.

Most data worth having does not sit still. Yesterday's weather, the current price of a euro, the earthquake that happened an hour ago, the roster of a repository that forty people are committing to. You cannot email yourself a copy of that and expect it to stay correct.

An **API** — Application Programming Interface — is how a program asks another program for data and gets a structured answer back. The useful word is *interface*: the part of a system you are allowed to touch. A microwave has an interface of buttons. You are not expected to open the case, and the manufacturer is free to change what is inside as long as the buttons keep working.

### The same data, two different doors

Open `weather.gov` in a browser and you get a page: fonts, a map, a navigation bar, an advert for the mobile app. That page is built for a person with eyes.

Open `api.weather.gov` and you get JSON. No fonts, no layout, no navigation — just the values, labeled. That is built for a program.

Same organization, same underlying data, two doors. The browser door is easy for you and miserable for your code, because getting numbers out of a web page means scraping HTML that breaks the next time a designer moves a `<div>`. The API door is the opposite: slightly more work to knock on, and then the answer arrives in a shape you can use.

This is why the JSON work from the last two weeks came first. **An API is a delivery mechanism; JSON is almost always what it delivers.** You already know what to do with the payload once it lands.

### Your first request, in three lines

Frankfurter publishes exchange rates from the European Central Bank. No signup, no key, no limits worth worrying about. Run this.

In [1]:
import requests

r = requests.get('https://api.frankfurter.app/latest',
                 params={'from': 'USD', 'to': 'EUR,GBP,JPY'},
                 timeout=10)

print('status:', r.status_code)
print(r.json())

status: 200
{'amount': 1.0, 'base': 'USD', 'date': '2026-10-07', 'rates': {'EUR': 0.89469, 'GBP': 0.75731, 'JPY': 158.23}}


That is the whole idea. You named a URL, you attached some parameters, and a server somewhere in Europe sent back today's rates as JSON. `r.json()` is the same `json.loads` from Monday's lecture, wrapped up for you.

Three details in that cell are not decoration:

- **`params=`** — you hand `requests` a dictionary and it builds the query string. Writing the URL by hand means escaping spaces and ampersands yourself, and getting it wrong.
- **`timeout=10`** — without it, a server that never answers hangs your notebook until you restart the kernel. There is no default timeout. Set one on every request you ever write.
- **`r.status_code`** — the server's report on whether it did what you asked. Look at it before you look at the body, every single time.

### Why we reach for an API instead of a file

**It is current.** The answer is generated when you ask. A file is a photograph of a moment that has passed.

**You ask for only what you need.** The weather archive holds every hour since 1940 for the whole planet. You want three days in Charlottesville, so that is what you request and that is what crosses the network.

**It reaches data you could never gather.** Nobody in this room is going to instrument the seismic network, or collect every exchange rate in Europe. Somebody already did, and they will answer the phone.

**There is no copy to maintain.** The moment you download `weather_2026.csv` you own a file that is quietly going stale, and a question about whether the one in your folder matches the one in your teammate's.

And the cost, which is the part people skip: **you are a guest on someone else's computer.** It can be slow. It can be down. It can rate-limit you, change its response shape, start charging, or shut off. Every one of those is a Tuesday, not a catastrophe — but your code has to expect them, which is the second half of today.

### Open APIs you can call right now

These need no signup and no key. Every one was tested the night before class.

| API | What it gives you | Typical use |
|---|---|---|
| **Open-Meteo** | Hourly and daily weather, 1940 to now | The midterm. Attendance against rain |
| **USGS Earthquakes** | Every recorded quake, as GeoJSON | Mapping, time series of events |
| **Frankfurter** | Daily exchange rates from the ECB | Converting prices to one currency |
| **GitHub REST** | Repositories, issues, commits | Measuring project activity |
| **Wikipedia REST** | Article summaries and metadata | Enriching a list of names or places |

Open-Meteo is the one that matters for you, because your midterm runs on it. The others are here so you can see that the shape of the work does not change from one API to the next.

The helper below reports what came back instead of assuming. Note that it prints failures rather than hiding them — that distinction becomes the whole of Build 2.

It also sends a **User-Agent**: a short string naming the program making the request. By default `requests` announces itself as `python-requests/2.x`, and some providers refuse that outright — Wikipedia answers it with a **403** and a link to their robot policy. Identifying yourself is not authentication and it is not a key. It is the network equivalent of saying who is at the door, and it costs you one line.

In [2]:
HEADERS = {'User-Agent': 'DS2002-UVA-student (course project)'}

def peek(label, url, params=None, headers=None):
    """Call an API and report what came back. Returns the response, or None."""
    try:
        r = requests.get(url, params=params,
                         headers={**HEADERS, **(headers or {})}, timeout=10)
    except requests.RequestException as err:
        print(f'{label:<14} NETWORK FAILED  {type(err).__name__}')
        return None
    ctype = r.headers.get('Content-Type', '').split(';')[0]
    print(f'{label:<14} {r.status_code}  {ctype}')
    return r


quakes = peek('USGS', 'https://earthquake.usgs.gov/fdsnws/event/1/query',
              {'format': 'geojson', 'starttime': '2026-09-29',
               'endtime': '2026-10-06', 'minmagnitude': 5})

repo = peek('GitHub', 'https://api.github.com/repos/pandas-dev/pandas')

wiki = peek('Wikipedia',
            'https://en.wikipedia.org/api/rest_v1/page/summary/University_of_Virginia')

USGS           200  application/json
GitHub         200  application/json
Wikipedia      200  application/json


Three `200`s. Now pull one useful fact out of each, which is where they stop looking alike.

In [3]:
print('earthquakes M5+:', len(quakes.json()['features']))
print('pandas stars   :', repo.json()['stargazers_count'])
print('UVA summary    :', wiki.json()['extract'][:90], '...')

earthquakes M5+: 40
pandas stars   : 49926
UVA summary    : The University of Virginia is a public research university located in Charlottesville City ...


Three APIs, three completely different response shapes, one `requests.get`.

The earthquakes arrive as **GeoJSON**, so the records sit in a list under `features` and the count is a `len()`. GitHub returns a single flat object, so the star count is one key lookup. Wikipedia returns a small object with the article text under `extract`.

That is the thing to take with you: **learning a new API is almost entirely learning its response shape.** The calling part — URL, parameters, timeout, status code — is the same every time. The reading part is the JSON work you have been doing for two weeks.

### Why some APIs want a key

An **API key** is a string that identifies you to the server. It is a nametag, not a password — it rarely unlocks anything secret. What it does is let the provider answer four questions:

1. **Who is this?** So they can contact you, or publish usage stats.
2. **How much have they used?** So they can cap you at 1,000 calls a day instead of guessing.
3. **Should they be billed?** Free tier, paid tier, or over the line.
4. **Can we cut them off?** One key can be revoked without affecting anyone else.

That is the entire reason keys exist. An open API like Open-Meteo skips all four by limiting everyone the same way.

Keys travel in one of two places. Some APIs want it in the query string, like `?api_key=abc123`. Most modern ones want it in a **header**, usually `Authorization: Bearer abc123`, which keeps it out of the URL and therefore out of server logs and browser history. The documentation will tell you which.

You do not need to sign up for anything to see what a missing key looks like. Run this — both of these are expected to fail, and the point is *how* they fail.

In [4]:
# An API that requires a key, called without one
owm = peek('OpenWeather', 'https://api.openweathermap.org/data/2.5/weather',
           {'q': 'Charlottesville'})
print('   ', owm.text[:110])
print()

# The same GitHub endpoint as before, but with a key that is not real
bad = peek('GitHub+badkey', 'https://api.github.com/user',
           headers={'Authorization': 'Bearer ghp_thisisnotarealtoken000000'})
print('   ', bad.text[:110])

OpenWeather    401  application/json
    {"cod":401, "message": "Invalid API key. Please see https://openweathermap.org/faq#error401 for more info."}

GitHub+badkey  401  application/json
    {
  "message": "Bad credentials",
  "documentation_url": "https://docs.github.com/rest",
  "status": "401"


**401 Unauthorized** both times, and in both cases the body says plainly what is wrong. That is the good case: a clear refusal you can read.

You will also meet **403 Forbidden**, which means the server knows who you are and you still may not have this. A key that is valid but out of quota usually returns 403 or 429, not 401.

Neither of these is worth retrying. No amount of waiting will make a missing key appear.

### A status of 200 does not mean you got data

This is the one that costs people an afternoon. The Census Bureau lets you make a small number of calls without a key, and when it decides you needed one, it does not return 401. It returns **200 OK** and an HTML error page.

In [5]:
census = peek('Census', 'https://api.census.gov/data/2022/acs/acs5',
              {'get': 'NAME,B01001_001E', 'for': 'state:51'})

print('first 60 characters:', census.text[:60])
print()
try:
    census.json()
except Exception as err:
    print('r.json() failed:', type(err).__name__, '-', err)

Census         200  text/html
first 60 characters: <html style="font-size: 14px;">

<head>
    <title>Missing K

r.json() failed: JSONDecodeError - Expecting value: line 1 column 1 (char 0)


Status 200. Content-Type `text/html`. And `r.json()` blows up with a `JSONDecodeError` complaining about line 1 column 1, which is the least helpful error message in this course.

That error does not mean the JSON is malformed. It means **the thing you got was never JSON at all** — it is a web page that starts with `<html>`, and `<` is not a legal way to begin a JSON document.

This is the habit worth taking from today: when `r.json()` fails, do not stare at the traceback. **Print `r.text[:200]` and look at what actually arrived.** Nine times out of ten it is an HTML error page, a login redirect, or a rate-limit notice, and your real problem is the request rather than the parsing.

### Rate limits, and how to see yours

Every API caps how often you may ask. Go over the cap and you get **429 Too Many Requests**. Polite providers tell you where you stand in the response headers before you get there.

In [6]:
rl = peek('GitHub', 'https://api.github.com/rate_limit')

core = rl.json()['resources']['core']
print('limit    :', core['limit'], 'requests per hour')
print('remaining:', core['remaining'])

print()
print('headers GitHub sent back:')
for h in ('X-RateLimit-Limit', 'X-RateLimit-Remaining', 'X-RateLimit-Reset'):
    print(f'  {h}: {rl.headers.get(h)}')

GitHub         200  application/json
limit    : 60 requests per hour
remaining: 59

headers GitHub sent back:
  X-RateLimit-Limit: 60
  X-RateLimit-Remaining: 59
  X-RateLimit-Reset: 1791403969


Sixty requests an hour without a key. With a key GitHub raises that to five thousand, which is the practical reason most people eventually get one.

Two rules that will keep you out of trouble all semester. **Ask once and keep the answer** — that is Build 4. And **never put a request inside a tight loop** without a pause; a `for` loop over 500 rows firing 500 requests as fast as Python can manage is indistinguishable from an attack, and you will be blocked.

### Where a key must never go

You push your work to GitHub. A key pasted into a notebook cell gets committed, and public repositories are scanned continuously by people harvesting credentials — a leaked key is typically found and used within minutes. Deleting it later does not help, because it stays in the commit history.

```python
KEY = 'a7f3c9e21b'        # never, not even for five minutes
```

In Colab, use the key icon in the left sidebar to store it as a secret, then read it:

```python
from google.colab import userdata
KEY = userdata.get('OPENWEATHER_KEY')
```

Running locally, set an environment variable and read it with `os.environ`:

```python
import os
KEY = os.environ['OPENWEATHER_KEY']
```

Either way the secret lives outside the file, the notebook is safe to commit, and a teammate runs the same code with their own key. Nothing today needs a key, but the midterm might, and this is the moment to learn the habit.

### What matters about an API, in one list

When you meet a new one, these are the things to find in the documentation before you write any code:

| | Why you care |
|---|---|
| **Base URL and endpoint** | Which question you are asking |
| **Parameters** | How you narrow it down |
| **Authentication** | Key or no key, query string or header |
| **Rate limit** | How often you may ask |
| **Response shape** | What one record looks like, and what is optional |
| **Terms of use** | What you are allowed to do with the data, and how to credit it |
| **Versioning** | The `/v1/` in the URL. A `/v2/` can change everything |

Together these are the **contract**. The provider promises to answer, in a documented shape, as long as you stay inside the limits. It is a real contract and it is also one they can rewrite, which is why the last thing you build today is code that notices when something has changed.

---

## Part 2 — Getting the data, and surviving the attempt

The rest of the studio is Open-Meteo, the API your midterm runs on.

### What a request actually consists of

```
https://archive-api.open-meteo.com/v1/archive ? latitude=38.03 & start_date=2026-10-03 & ...
|------------------ base URL -------------------|  |--------------- query parameters ---------------|
```

You send that; the server sends back a **status code** and a **body**. The status code is the first thing to look at, every time.

| Code | Meaning | What you should do |
|---|---|---|
| 200 | Fine | Parse the body |
| 400 | Your request was malformed | Fix your parameters. Retrying will not help |
| 404 | Wrong URL | Fix the URL. Retrying will not help |
| 429 | You are asking too fast | Slow down and retry |
| 500–504 | Their server had a problem | Wait and retry |

That last column is the whole design of the retry logic you will write in a few minutes: retry the server's problems, never retry your own mistakes.

### Build 1 — see the URL before you send it

`requests` builds the query string from a dict for you. Print it once so the mapping between your dict and the actual URL stops being abstract.

Note how the dates are built. The archive endpoint only holds the **past**, and the most recent day or two is still settling, so asking for today or tomorrow is an error rather than an empty answer. Computing the window backwards from `date.today()` means this notebook still runs in November, which a pair of dates typed in by hand would not.

In [7]:
import requests, pandas as pd
from urllib.parse import urlencode
from datetime import date, timedelta

BASE = 'https://archive-api.open-meteo.com/v1/archive'

END = date.today() - timedelta(days=3)     # let the archive settle
START = END - timedelta(days=2)            # a three-day window

params = {
    'latitude': 38.03, 'longitude': -78.51,          # Charlottesville
    'start_date': START.isoformat(), 'end_date': END.isoformat(),
    'hourly': 'temperature_2m,precipitation',
    'timezone': 'America/New_York',
}

print(BASE + '?' + urlencode(params))

https://archive-api.open-meteo.com/v1/archive?latitude=38.03&longitude=-78.51&start_date=2026-10-02&end_date=2026-10-04&hourly=temperature_2m%2Cprecipitation&timezone=America%2FNew_York


Now send it. Note the three things this cell does that a careless version would skip: it sets a **timeout**, it looks at the **status code**, and it prints the top-level shape of the response instead of assuming it knows what came back.

In [8]:
resp = requests.get(BASE, params=params, timeout=15)

print('status:', resp.status_code)
print('keys:', list(resp.json().keys()))
print('units:', resp.json()['hourly_units'])

status: 200
keys: ['latitude', 'longitude', 'generationtime_ms', 'utc_offset_seconds', 'timezone', 'timezone_abbreviation', 'elevation', 'hourly_units', 'hourly']
units: {'time': 'iso8601', 'temperature_2m': '°C', 'precipitation': 'mm'}


Open-Meteo hands back parallel lists — one list of timestamps, one list per variable. That shape drops straight into a DataFrame.

In [9]:
hourly = pd.DataFrame(resp.json()['hourly'])
hourly['time'] = pd.to_datetime(hourly['time'])
print(hourly.shape)
hourly.head()

(72, 3)


,time,temperature_2m,precipitation
0,2026-10-02 00:00:00,21.7,0.0
1,2026-10-02 01:00:00,21.1,0.0
2,2026-10-02 02:00:00,17.9,0.0
3,2026-10-02 03:00:00,17.3,0.0
4,2026-10-02 04:00:00,16.4,0.0


### Build 2 — what failure actually looks like

Send a request that is wrong on purpose: ask the *archive* for a date that has not happened yet. This is not a contrived mistake — it is the single most common way people break this endpoint, usually by reaching for the forecast and typing the archive URL.

Read the status code, then read the body. The body tells you exactly what you got wrong, and most people never look at it.

In [10]:
future = date.today() + timedelta(days=10)
bad = dict(params, start_date=future.isoformat(),
           end_date=(future + timedelta(days=2)).isoformat())

r = requests.get(BASE, params=bad, timeout=15)

print('status:', r.status_code)
print('body:', r.text[:300])

status: 400
body: {"error":true,"reason":"Parameter 'start_date' is out of allowed range from 1940-01-01 to 2026-10-07"}


**400**, and the body names the parameter and the range it will accept:

```
{"error":true,"reason":"Parameter 'start_date' is out of allowed range from 1940-01-01 to ..."}
```

Two things to take from that. The server told you precisely what to fix, which is more than most error messages manage. And **retrying this would be pointless** — it will be just as out of range in thirty seconds. That distinction is the whole design of the retry logic you are about to write.

Now the trap. This is the most damaging pattern in data engineering, and it looks responsible:

```python
try:
    wx = fetch_weather(date)
except Exception:
    wx = pd.DataFrame()   # <- never do this
```

An empty frame from a failed call and an empty frame from a genuinely dry day are indistinguishable downstream. Your join silently drops rows, your chart shows no rain, and you present a conclusion built on a network error.

A failure must either raise, or be recorded as a failure. It must never disguise itself as data.

Here is the version that behaves. It is four lines, and the whole point is the third one.

In [11]:
def fetch_json(url, params, timeout=15):
    """Return parsed JSON, or raise an error that says what went wrong."""
    r = requests.get(url, params=params, timeout=timeout)
    if r.status_code != 200:
        raise RuntimeError(f'{url} returned {r.status_code}: {r.text[:200]}')
    return r.json()

# The good request:
print(list(fetch_json(BASE, params).keys()))

print()
# The bad one, raising something you can actually act on:
try:
    fetch_json(BASE, bad)
except Exception as e:
    print('raised:', e)

['latitude', 'longitude', 'generationtime_ms', 'utc_offset_seconds', 'timezone', 'timezone_abbreviation', 'elevation', 'hourly_units', 'hourly']

raised: https://archive-api.open-meteo.com/v1/archive returned 400: {"reason":"Parameter 'start_date' is out of allowed range from 1940-01-01 to 2026-10-07","error":true}


Read what that error message carries: the URL, the status code, and the first 200 characters of the body. Six weeks from now, when this fires inside a pipeline at two in the morning, those three things are the difference between a fix and an afternoon.

Compare it to the alternatives. `pass` gives you `None` and a `TypeError` three cells later. `raise RuntimeError('request failed')` tells you nothing you did not already know. `r.raise_for_status()` is a real option and better than both, but it throws away the body — and on this API the body is where the server told you *which parameter* was wrong.

**Why `r.text[:200]` and not all of it.** A failing endpoint sometimes returns a megabyte of HTML. Truncating keeps your traceback readable while still showing you whether you got JSON, an error page, or a login redirect.

### Build 3 — retry the failures worth retrying

A 500 or a timeout usually means "try again in a moment." A 400 means "you are wrong, and you will still be wrong in a moment."

Back off between attempts instead of hammering: wait 1 second, then 2, then 4. If you retry instantly in a loop you look like an attack and get rate-limited.

First, prove the logic without the network. `flaky()` fails twice and then succeeds, so the test is deterministic.

Watch the printed delays when you run it.

In [12]:
import time

attempts = {'n': 0}

def flaky():
    """Fails the first two times it is called, then succeeds."""
    attempts['n'] += 1
    if attempts['n'] < 3:
        raise TimeoutError(f"simulated timeout (call {attempts['n']})")
    return {'ok': True, 'calls_needed': attempts['n']}

def with_retries(fn, tries=4, base_delay=0.1):
    """Call fn(), retrying on failure with an increasing delay.
    Re-raise if every attempt fails."""
    for attempt in range(tries):
        try:
            return fn()
        except Exception as e:
            if attempt == tries - 1:   # last go: let it out
                raise
            delay = base_delay * (2 ** attempt)
            print(f'attempt {attempt + 1} failed ({e}); retrying in {delay}s')
            time.sleep(delay)

print(with_retries(flaky))

attempt 1 failed (simulated timeout (call 1)); retrying in 0.1s
attempt 2 failed (simulated timeout (call 2)); retrying in 0.2s
{'ok': True, 'calls_needed': 3}


`0.1s`, then `0.2s`, then it succeeds on the third call. That doubling is **exponential backoff**, and `base_delay * (2 ** attempt)` is the whole of it.

Three details that are easy to get wrong:

- **`return fn()` sits inside the `try`.** A success leaves the loop immediately. There is no "did it work" flag to keep track of.
- **The last attempt re-raises instead of swallowing.** Without that check the function falls off the end of the loop and quietly returns `None`, which is the empty-DataFrame trap wearing a different hat.
- **It sleeps *between* attempts, not after the last one.** Sleeping after a failure you are about to give up on just makes the user wait.

Testing against `flaky()` rather than the network is deliberate. The retry logic is the part you need to be sure about, and a real server will not fail on command.

Now the same idea wired into a real fetch. Retry on 429 and 5xx and on network errors. Do **not** retry on 400 or 404 — raise immediately, because retrying a bad request just wastes everyone's time.

In [13]:
RETRY_STATUSES = {429, 500, 502, 503, 504}

def fetch_with_retries(url, params, tries=4, base_delay=0.5):
    last = None
    for attempt in range(tries):
        try:
            r = requests.get(url, params=params, timeout=15)
            if r.status_code == 200:
                return r.json()
            if r.status_code not in RETRY_STATUSES:
                raise RuntimeError(
                    f'not retryable: {r.status_code} {r.text[:200]}')
            last = f'status {r.status_code}'
        except requests.RequestException as e:
            last = repr(e)
        delay = base_delay * (2 ** attempt)
        print(f'attempt {attempt + 1} failed ({last}); retrying in {delay}s')
        time.sleep(delay)
    raise RuntimeError(f'gave up after {tries} attempts: {last}')

try:
    data = fetch_with_retries(BASE, params)
    print('got', len(data['hourly']['time']), 'hourly readings')
except Exception as e:
    print('no network in this room:', e)
    data = None

got 72 hourly readings


On a working network this prints nothing but the success line, because the first attempt returns 200.

The shape of this function is the part worth keeping. There are **three** different outcomes and they are handled in three different places:

| Outcome | What happens | Why |
|---|---|---|
| `200` | `return r.json()` | Done. Leave immediately |
| 400, 404, 401 | `raise` on the spot | Your bug. Waiting changes nothing |
| 429, 5xx, network error | record it, sleep, loop | Their problem. It may clear |

**`except requests.RequestException`** catches the case where there is no response at all — DNS failure, refused connection, the timeout expiring. Those never reach a status code, so a function that only checks `r.status_code` misses them entirely.

**The final `raise` matters.** If all four attempts fail, the function raises rather than returning `None`. A caller that gets an exception knows something went wrong. A caller that gets `None` finds out later, somewhere less convenient.

The `try` around the call at the bottom is there only so the rest of the notebook still runs if the room's wifi is out. Note that it sets `data = None` and **says so on screen** — it is reporting the failure, not disguising it as data.

### Build 4 — stop asking the same question twice *(drop this one if time is short)*

Re-running a notebook cell should not mean re-hitting somebody's server. A small cache makes your notebook faster, makes your results stable while you work, and keeps you off the rate limit.

The second call below never reaches the network.

In [14]:
_cache = {}

def cache_key(params):
    return tuple(sorted(params.items()))

def cached_fetch(url, params):
    key = cache_key(params)
    if key in _cache:
        print('cache hit')
        return _cache[key]
    value = fetch_with_retries(url, params)
    _cache[key] = value
    return value

cached_fetch(BASE, params)   # hits the network
cached_fetch(BASE, params)   # prints 'cache hit'
print('cached requests:', len(_cache))

cache hit
cached requests: 1


One entry in the cache after two calls.

`cache_key` is the only interesting line. A dictionary cannot be a dictionary key, because it is mutable and therefore unhashable, so `tuple(sorted(params.items()))` turns it into something that can be. **The `sorted` is load-bearing**: without it, the same parameters written in a different order would produce a different key and you would fetch the same data twice.

This cache lives in memory and dies with the kernel. That is the right size for a studio. For the midterm, writing the response to a `.json` file and checking whether that file exists gives you the same benefit across restarts — and makes your results reproducible, since you can hand a teammate the file instead of hoping the API returns the same thing tomorrow.

### Build 5 — hourly readings into the shape you can join

Sales data is daily. Weather came back hourly. Somebody has to reshape, and the choice of aggregation is a real decision: rain **sums** over a day, temperature **maxes**.

One row per calendar day, with total precipitation and max temperature.

In [15]:
if data:
    hourly = pd.DataFrame(data['hourly'])
else:                                  # keeps the notebook usable offline
    hourly = pd.DataFrame({
        'time': pd.date_range(START, periods=72, freq='h').astype(str),
        'temperature_2m': [12 + (i % 12) for i in range(72)],
        'precipitation': [0.0] * 30 + [1.4] * 6 + [0.0] * 36,
    })

hourly['time'] = pd.to_datetime(hourly['time'])

daily = (hourly.groupby(hourly['time'].dt.date)
         .agg(precipitation_total=('precipitation', 'sum'),
              temp_max=('temperature_2m', 'max'))
         .reset_index()
         .rename(columns={'time': 'date'}))
daily

,date,precipitation_total,temp_max
0,2026-10-02,1.6,31.4
1,2026-10-03,14.7,22.0
2,2026-10-04,1.8,19.1


Three days in, three rows out, which is the check worth making before you trust any reshape.

The named aggregation syntax — `precipitation_total=('precipitation', 'sum')` — does two jobs at once: it picks the column and names the result. That is why no second `rename` is needed for those two columns.

**The aggregation choice is the real decision here, not the syntax.** Rain accumulates, so it sums. Temperature is a reading at a moment, so adding up 24 of them is meaningless and you take the max instead. Get this backwards and the code still runs and the numbers are still wrong, which is the worst combination available.

`groupby(hourly['time'].dt.date)` groups on the calendar date pulled out of the timestamp. Grouping on `time` itself would give you one group per *hour* — 72 groups of one row each, and no error to warn you.

One more question off the same frame: which single hour had the most rain?

In [16]:
wettest = hourly.loc[hourly['precipitation'].idxmax()]
print('wettest hour:', wettest['time'], '|', wettest['precipitation'], 'mm')

wettest hour: 2026-10-03 15:00:00 | 3.3 mm


`idxmax()` returns the *index label* of the largest value, and `.loc` turns that label into the whole row. That two-step is the pattern any time you want the record holding an extreme rather than the extreme itself — `max()` on its own gives you the millimetres and throws away the timestamp, which is usually the part you were after.

It is also the reason to keep the hourly table after building the daily one. The daily roll-up cannot answer this question: the hour it happened has already been summed away.

### Cite what you pulled

Every project in this course that uses an API has to say where the data came from. Not as a formality — six weeks from now you will need to know whether a number came from the archive endpoint or the forecast endpoint, and which coordinates you used.

In [17]:
from datetime import datetime

print('Source:  Open-Meteo historical archive API')
print('Endpoint:', BASE)
print('Params: ', params)
print('Pulled: ', datetime.now().strftime('%Y-%m-%d %H:%M'))

Source:  Open-Meteo historical archive API
Endpoint: https://archive-api.open-meteo.com/v1/archive
Params:  {'latitude': 38.03, 'longitude': -78.51, 'start_date': '2026-10-02', 'end_date': '2026-10-04', 'hourly': 'temperature_2m,precipitation', 'timezone': 'America/New_York'}
Pulled:  2026-10-07 19:12


---

## Checkpoint (participation)

Report the daily precipitation total your reshape produced, and what happened when you sent the deliberately bad request.

Work in groups if you wish, then fill in the cell below **yourself**. Paste the printed output (or a screenshot of it) into this week's **Studio Checkpoint** in Canvas by **Thursday 11:59pm ET**. One submission per person, not per group.

In [18]:
# Checkpoint
daily_rows = len(daily)

wettest_row = daily.loc[daily['precipitation_total'].idxmax()]
wettest_day = str(wettest_row['date'])

# r is the response from the deliberately bad request in Build 2
bad_request_status = r.status_code

retry_rule = (
    '400 Bad Request: the parameters are invalid. '
    'Retrying the same request will not fix them.'
)

print('daily rows:', daily_rows)
print('wettest day:', wettest_day)
print(
    'wettest day precipitation total:',
    wettest_row['precipitation_total'],
    'mm'
)
print('bad request returned:', bad_request_status)
print('would not retry:', retry_rule)

print('\nDaily precipitation totals:')
print(daily[['date', 'precipitation_total']].to_string(index=False))

if data is None:
    print('\nNOTE: These totals use the notebook’s synthetic offline data.')

daily rows: 3
wettest day: 2026-10-03
wettest day precipitation total: 14.7 mm
bad request returned: 400
would not retry: 400 Bad Request: the parameters are invalid. Retrying the same request will not fix them.

Daily precipitation totals:
      date  precipitation_total
2026-10-02                  1.6
2026-10-03                 14.7
2026-10-04                  1.8
